File: backend/lib/schemas/prescription.ts
import { z } from 'zod'

// Dosage schema for medication grid
export const dosageSchema = z.object({
  amount: z.string().min(1, 'Dosage amount is required'),
  beforeMeal: z.boolean(),
}).optional()

// Medication schema for prescription creation
export const medicationSchema = z.object({
  rowNumber: z.number().int().positive(),
  medicineName: z.string().min(1, 'Medicine name is required'),
  medicineNameKhmer: z.string().optional(),
  imageUrl: z.string().url().optional(),
  morningDosage: dosageSchema,
  daytimeDosage: dosageSchema,
  nightDosage: dosageSchema,
  frequency: z.string().optional(),
  timing: z.string().optional(),
}).refine(
  (data) => data.morningDosage || data.daytimeDosage || data.nightDosage,
  {
    message: 'At least one dosage (morning, daytime, or night) must be specified',
  }
)

// Prescription creation schema
export const prescriptionCreateSchema = z.object({
  patientId: z.string().uuid('Invalid patient ID'),
  patientName: z.string().min(1, 'Patient name is required'),
  patientGender: z.enum(['MALE', 'FEMALE', 'OTHER']),
  patientAge: z.number().int().min(0).max(150),
  symptoms: z.string().min(1, 'Symptoms are required'),
  medications: z.array(medicationSchema).min(1, 'At least one medication is required'),
  status: z.enum(['DRAFT', 'ACTIVE']).default('DRAFT'),
  isUrgent: z.boolean().default(false),
  urgentReason: z.string().optional(),
}).refine(
  (data) => {
    // If urgent, reason is required
    if (data.isUrgent && !data.urgentReason) {
      return false
    }
    return true
  },
  {
    message: 'Urgent reason is required when prescription is marked as urgent',
    path: ['urgentReason'],
  }
)

// Prescription update schema
export const prescriptionUpdateSchema = z.object({
  status: z.enum(['DRAFT', 'ACTIVE', 'PAUSED', 'INACTIVE']).optional(),
  medications: z.array(medicationSchema).optional(),
  isUrgent: z.boolean().optional(),
  urgentReason: z.string().optional(),
  symptoms: z.string().optional(),
}).refine(
  (data) => {
    // If urgent, reason is required
    if (data.isUrgent && !data.urgentReason) {
      return false
    }
    return true
  },
  {
    message: 'Urgent reason is required when prescription is marked as urgent',
    path: ['urgentReason'],
  }
)

// Prescription confirm schema (patient accepting prescription)
export const prescriptionConfirmSchema = z.object({
  confirmed: z.boolean(),
})

// Prescription retake schema (patient requesting revision)
export const prescriptionRetakeSchema = z.object({
  reason: z.string().min(1, 'Reason for retake is required'),
})

// Prescription query filters
export const prescriptionQuerySchema = z.object({
  patientId: z.string().uuid().optional(),
  doctorId: z.string().uuid().optional(),
  status: z.enum(['DRAFT', 'ACTIVE', 'PAUSED', 'INACTIVE']).optional(),
  page: z.string().transform(Number).pipe(z.number().int().positive()).default('1'),
  limit: z.string().transform(Number).pipe(z.number().int().min(1).max(100)).default('50'),
})

// Export types
export type PrescriptionCreate = z.infer<typeof prescriptionCreateSchema>
export type PrescriptionUpdate = z.infer<typeof prescriptionUpdateSchema>
export type PrescriptionConfirm = z.infer<typeof prescriptionConfirmSchema>
export type PrescriptionRetake = z.infer<typeof prescriptionRetakeSchema>
export type PrescriptionQuery = z.infer<typeof prescriptionQuerySchema>
export type Medication = z.infer<typeof medicationSchema>
export type Dosage = z.infer<typeof dosageSchema>

File: mobile_app/lib/models/prescription_model/prescription.dart
class Prescription {
  final String? id;
  final String patientId;
  final String? doctorId;
  final String patientName;
  final String patientGender;
  final int patientAge;
  final String symptoms;
  final String? diagnosis;
  final String? clinicalNote;
  final String? followUpDate;
  final String status;
  final List<PrescriptionMedication> medications;
  final int currentVersion;
  final bool isUrgent;
  final String? urgentReason;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Prescription({
    this.id,
    required this.patientId,
    this.doctorId,
    required this.patientName,
    required this.patientGender,
    required this.patientAge,
    required this.symptoms,
    this.diagnosis,
    this.clinicalNote,
    this.followUpDate,
    required this.status,
    required this.medications,
    this.currentVersion = 1,
    this.isUrgent = false,
    this.urgentReason,
    this.createdAt,
    this.updatedAt,
  });

  factory Prescription.fromJson(Map<String, dynamic> json) {
    return Prescription(
      id: json['id'] as String?,
      patientId: json['patientId'] as String,
      doctorId: json['doctorId'] as String?,
      patientName: json['patientName'] as String? ?? '',
      patientGender: json['patientGender'] as String? ?? 'OTHER',
      patientAge: json['patientAge'] as int? ?? 0,
      symptoms: json['symptoms'] as String? ?? '',
      diagnosis: json['diagnosis'] as String?,
      clinicalNote: json['clinicalNote'] as String?,
      followUpDate: json['followUpDate'] as String?,
      status: json['status'] as String? ?? 'DRAFT',
      medications: json['medications'] != null
          ? (json['medications'] as List)
              .map((m) => PrescriptionMedication.fromJson(m))
              .toList()
          : [],
      currentVersion: json['currentVersion'] as int? ?? 1,
      isUrgent: json['isUrgent'] as bool? ?? false,
      urgentReason: json['urgentReason'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  /// Converts to JSON for reading/display purposes.
  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'patientId': patientId,
      if (doctorId != null) 'doctorId': doctorId,
      'patientName': patientName,
      'patientGender': patientGender,
      'patientAge': patientAge,
      'symptoms': symptoms,
      if (diagnosis != null) 'diagnosis': diagnosis,
      if (clinicalNote != null) 'clinicalNote': clinicalNote,
      if (followUpDate != null) 'followUpDate': followUpDate,
      'medications': medications.map((m) => m.toJson()).toList(),
      'isUrgent': isUrgent,
      if (urgentReason != null) 'urgentReason': urgentReason,
    };
  }
}

class PrescriptionMedication {
  final String? id;
  final String? prescriptionId;
  final int rowNumber;
  final String medicineName;
  final String? medicineNameKhmer;
  final String? medicineType;
  final String? unit;
  final double? dosageAmount;
  final String? description;
  final String? additionalNote;
  final String? imageUrl;
  final String? frequency;
  final int? duration;
  final String? timing;
  final bool isPRN;
  final bool beforeMeal;
  final Map<String, dynamic>? morningDosage;
  final Map<String, dynamic>? daytimeDosage;
  final Map<String, dynamic>? nightDosage;
  final String? createdBy;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  PrescriptionMedication({
    this.id,
    this.prescriptionId,
    required this.rowNumber,
    required this.medicineName,
    this.medicineNameKhmer,
    this.medicineType,
    this.unit,
    this.dosageAmount,
    this.description,
    this.additionalNote,
    this.imageUrl,
    this.frequency,
    this.duration,
    this.timing,
    this.isPRN = false,
    this.beforeMeal = false,
    this.morningDosage,
    this.daytimeDosage,
    this.nightDosage,
    this.createdBy,
    this.createdAt,
    this.updatedAt,
  });

  factory PrescriptionMedication.fromJson(Map<String, dynamic> json) {
    return PrescriptionMedication(
      id: json['id'] as String?,
      prescriptionId: json['prescriptionId'] as String?,
      rowNumber: json['rowNumber'] as int? ?? 0,
      medicineName: json['medicineName'] as String? ?? '',
      medicineNameKhmer: json['medicineNameKhmer'] as String?,
      medicineType: json['medicineType'] as String?,
      unit: json['unit'] as String?,
      dosageAmount: json['dosageAmount'] != null
          ? (json['dosageAmount'] as num).toDouble()
          : null,
      description: json['description'] as String?,
      additionalNote: json['additionalNote'] as String?,
      imageUrl: json['imageUrl'] as String?,
      frequency: json['frequency'] as String?,
      duration: json['duration'] as int?,
      timing: json['timing'] as String?,
      isPRN: json['isPRN'] as bool? ?? false,
      beforeMeal: json['beforeMeal'] as bool? ?? false,
      morningDosage: json['morningDosage'] != null
          ? Map<String, dynamic>.from(json['morningDosage'] as Map)
          : null,
      daytimeDosage: json['daytimeDosage'] != null
          ? Map<String, dynamic>.from(json['daytimeDosage'] as Map)
          : null,
      nightDosage: json['nightDosage'] != null
          ? Map<String, dynamic>.from(json['nightDosage'] as Map)
          : null,
      createdBy: json['createdBy'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      if (prescriptionId != null) 'prescriptionId': prescriptionId,
      'rowNumber': rowNumber,
      'medicineName': medicineName,
      if (medicineNameKhmer != null) 'medicineNameKhmer': medicineNameKhmer,
      if (medicineType != null) 'medicineType': medicineType,
      if (unit != null) 'unit': unit,
      if (dosageAmount != null) 'dosageAmount': dosageAmount,
      if (description != null) 'description': description,
      if (additionalNote != null) 'additionalNote': additionalNote,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (frequency != null) 'frequency': frequency,
      if (duration != null) 'durationDays': duration,
      'isPRN': isPRN,
      'beforeMeal': beforeMeal,
      if (morningDosage != null) 'morningDosage': morningDosage,
      if (daytimeDosage != null) 'daytimeDosage': daytimeDosage,
      if (nightDosage != null) 'nightDosage': nightDosage,
    };
  }
}
